README.md
[![Build Status](https://img.shields.io/github/workflow/status/xizzhu/Logger/Android%20CI/master?color=blue)](https://github.com/xizzhu/Logger/actions)
[![API](https://img.shields.io/badge/API-21%2B-green.svg?style=flat)](https://developer.android.com/about/versions/android-5.0.html)
[![GitHub license](https://img.shields.io/badge/license-Apache%20License%202.0-blue.svg?style=flat)](https://www.apache.org/licenses/LICENSE-2.0)
[![JitPack](https://img.shields.io/jitpack/v/github/xizzhu/Logger.svg)](https://jitpack.io/#me.xizzhu/logger)

Logger
======

Yet another simple logger for your awesome Android application, written in Kotlin.

Download
--------
* Gradle: Add the following to your `build.gradle`:
```gradle
repositories {
    maven { url "https://jitpack.io" }
}

dependencies {
    implementation 'me.xizzhu:logger:$latest_version'
}
```

Usage
-----
### Initialize Logger
```kotlin
// Write logs to LogCat
Log.addLogger(LogcatLogger())

// Write logs to the specified file
Log.addLogger(FileLogger(logFile))

// Write logs to Firebase Crashlytics
Log.addLogger(CrashlyticsLogger())
```

#### Specify Log Level
When creating `Logger`s, the app can specify a log level, e.g.:
```kotlin
// Write logs to LogCat, only if they are equal or higher than DEBUG level
Log.addLogger(LogcatLogger(level = Log.DEBUG))
```

### Write Logs
To write logs, you can e.g.:
```kotlin
// Write logs in VERBOSE level
Log.v(tag, msg)

// Write logs with an exception in VERBOSE level
Log.v(tag, msg, exception)
```

If no `Logger` has been initialized, calling `Log.v()` and other logging methods are no-op.

License
-------
    Copyright (C) 2021 Xizhi Zhu

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.md
CHANGELOG
---------

#### 0.7.0
- Changes:
  - Lower Minimum SDK Version to 19 (Android 4.4)

#### 0.6.1
- Changes:
  - Allow app to provide FirebaseCrashlytics instance

#### 0.6.0
- Changes:
  - Use GitHub as CI
  - Artifact group ID changed to me.xizzhu
- Updated dependencies:
  - Gradle to 7.0.2
  - Java and Kotlin JVM Target to 11
  - Android Gradle plugin to 7.0.3
  - Android Build SDK to 31.0.0, Compile and Target to 31
  - Kotlin to 1.5.31
  - Firebase Crashlytics to 18.2.3
  - Mockk to 1.12.0

#### 0.5.0
- New features:
  - Support log level
- Changes:
  - Simplify Logger interface by merging the two log() functions
  - Use mockk instead of Mockito
  - Remove jcenter() from repository dependencies
- Updated dependencies:
  - Gradle to 6.9
  - Android Gradle plugin to 4.2.1
  - Kotlin to 1.5.10
  - AndroidX Annotation to 1.2.0
  - Firebase Crashlytics to 18.0.1 

#### 0.4.1
- Bug fixes:
  - Should do `executor.execute()` instead of `run()` (#4)

#### 0.4.0
- New features:
  - Add CrashlyticsLogger
- Updated dependencies:
  - Gradle to 6.7.1
  - Android Gradle plugin to 4.1.1
  - Android build tool to 30.0.3
  - Kotlin to 1.4.21
  - Mockito to 3.6.28

#### 0.3.0
- Bug fixes:
  - Use internal file dir for sample app
  - Use CopyOnWriteArrayList to store list of loggers
- Updated dependencies:
  - Gradle to 6.6.1
  - Android Gradle plugin to 4.0.1
  - Android build tool to 30.0.2
  - Compile and target SDK to 30
  - Kotlin to 1.4.0
  - Mockito to 3.5.9

#### 0.2.0
- Used Executor instead of coroutine for FileLogger
- Updated dependencies:
  - Gradle to 6.2.2
  - Android Gradle plugin to 3.6.1
  - Android build tool to 29.0.3
  - Kotlin to 1.3.70
  - Mockito to 3.3.0

#### 0.1.5
- Updated dependencies:
  - Gradle to 6.0.1
  - Android Gradle plugin to 3.5.3
  - Kotlin to 1.3.61
  - Kotlin Coroutine to 1.3.1
  - Mockito to 3.2.0

#### 0.1.4
- Updated dependencies:
  - Kotlin Coroutine to 1.3.0

#### 0.1.3
- Updated dependencies:
  - Gradle to 5.6
  - Android Gradle plugin to 3.5.0
  - Android build tool to 29.0.2
  - Kotlin to 1.3.50
  - Mockito to 3.0.0

#### 0.1.2
- Updated dependencies:
  - Gradle to 5.5
  - Kotlin to 1.3.41
  - Kotlin Coroutine to 1.2.2

#### 0.1.1
- Bumped Kotlin to 1.3.40

#### 0.1.0
- Initial release


lib/src/test/kotlin/android/util/Log.kt
/*
 * Copyright (C) 2021 Xizhi Zhu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.util

class Log {
    companion object {
        @JvmStatic
        fun getStackTraceString(tr: Throwable): String = tr.message ?: "Unknown"
    }
}


lib/src/test/kotlin/me/xizzhu/android/logger/LoggerTest.kt
/*
 * Copyright (C) 2021 Xizhi Zhu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xizzhu.android.logger

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import kotlin.test.BeforeTest
import kotlin.test.Test

class LoggerTest {
    private lateinit var logger: Logger

    private val tag = "TAG"
    private val msg = "msg"
    private val exception = RuntimeException("Random exception")

    @BeforeTest
    fun setup() {
        logger = mockk()
        every { logger.log(any(), any(), any(), any()) } returns Unit
        every { logger.level } returns Log.VERBOSE
    }

    @Test
    fun testNoLogger() {
        Log.v(tag, msg)
        Log.d(tag, msg)
        Log.i(tag, msg)
        Log.w(tag, msg)
        Log.e(tag, msg)
        Log.f(tag, msg)

        verify(exactly = 0) {
            logger.log(any(), any(), any(), any())
        }
    }

    @Test
    fun testNoLoggerWithException() {
        Log.v(tag, msg, exception)
        Log.d(tag, msg, exception)
        Log.i(tag, msg, exception)
        Log.w(tag, msg, exception)
        Log.e(tag, msg, exception)
        Log.f(tag, msg, exception)

        verify(exactly = 0) {
            logger.log(any(), any(), any(), any())
        }
    }

    @Test
    fun testAddThenRemoveLogger() {
        Log.addLogger(logger)
        Log.removeLogger(logger)

        Log.v(tag, msg)
        Log.d(tag, msg)
        Log.i(tag, msg)
        Log.w(tag, msg)
        Log.e(tag, msg)
        Log.f(tag, msg)

        verify(exactly = 0) {
            logger.log(any(), any(), any(), any())
        }
    }

    @Test
    fun testAddThenRemoveLoggerWithException() {
        Log.addLogger(logger)
        Log.removeLogger(logger)

        Log.v(tag, msg, exception)
        Log.d(tag, msg, exception)
        Log.i(tag, msg, exception)
        Log.w(tag, msg, exception)
        Log.e(tag, msg, exception)
        Log.f(tag, msg, exception)

        verify(exactly = 0) {
            logger.log(any(), any(), any(), any())
        }
    }

    @Test
    fun testAddLoggerTwice() {
        Log.addLogger(logger)
        Log.addLogger(logger)

        Log.v(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.VERBOSE, tag, msg, null)
        }

        Log.d(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.DEBUG, tag, msg, null)
        }

        Log.i(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.INFO, tag, msg, null)
        }

        Log.w(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.WARN, tag, msg, null)
        }

        Log.e(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.ERROR, tag, msg, null)
        }

        Log.f(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.FATAL, tag, msg, null)
        }

        Log.removeLogger(logger)
    }

    @Test
    fun testAddLoggerTwiceWithException() {
        Log.addLogger(logger)
        Log.addLogger(logger)

        Log.v(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.VERBOSE, tag, msg, exception)
        }

        Log.d(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.DEBUG, tag, msg, exception)
        }

        Log.i(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.INFO, tag, msg, exception)
        }

        Log.w(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.WARN, tag, msg, exception)
        }

        Log.e(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.ERROR, tag, msg, exception)
        }

        Log.f(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.FATAL, tag, msg, exception)
        }

        Log.removeLogger(logger)
    }

    @Test
    fun testLog() {
        Log.addLogger(logger)

        Log.v(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.VERBOSE, tag, msg, null)
        }

        Log.d(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.DEBUG, tag, msg, null)
        }

        Log.i(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.INFO, tag, msg, null)
        }

        Log.w(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.WARN, tag, msg, null)
        }

        Log.e(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.ERROR, tag, msg, null)
        }

        Log.f(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.FATAL, tag, msg, null)
        }

        Log.removeLogger(logger)
    }

    @Test
    fun testLogWithException() {
        Log.addLogger(logger)

        Log.v(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.VERBOSE, tag, msg, exception)
        }

        Log.d(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.DEBUG, tag, msg, exception)
        }

        Log.i(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.INFO, tag, msg, exception)
        }

        Log.w(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.WARN, tag, msg, exception)
        }

        Log.e(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.ERROR, tag, msg, exception)
        }

        Log.f(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.FATAL, tag, msg, exception)
        }

        Log.removeLogger(logger)
    }

    @Test
    fun testLogLevel() {
        every { logger.level } returns Log.INFO
        Log.addLogger(logger)

        Log.v(tag, msg)
        verify(exactly = 0) {
            logger.log(Log.VERBOSE, any(), any(), any())
        }

        Log.d(tag, msg)
        verify(exactly = 0) {
            logger.log(Log.DEBUG, any(), any(), any())
        }

        Log.i(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.INFO, tag, msg, null)
        }

        Log.w(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.WARN, tag, msg, null)
        }

        Log.e(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.ERROR, tag, msg, null)
        }

        Log.f(tag, msg)
        verify(exactly = 1) {
            logger.log(Log.FATAL, tag, msg, null)
        }

        Log.removeLogger(logger)
    }

    @Test
    fun testLogLevelWithException() {
        every { logger.level } returns Log.WARN
        Log.addLogger(logger)

        Log.v(tag, msg, exception)
        verify(exactly = 0) {
            logger.log(Log.VERBOSE, any(), any(), any())
        }

        Log.d(tag, msg, exception)
        verify(exactly = 0) {
            logger.log(Log.DEBUG, any(), any(), any())
        }

        Log.i(tag, msg, exception)
        verify(exactly = 0) {
            logger.log(Log.INFO, any(), any(), any())
        }

        Log.w(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.WARN, tag, msg, exception)
        }

        Log.e(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.ERROR, tag, msg, exception)
        }

        Log.f(tag, msg, exception)
        verify(exactly = 1) {
            logger.log(Log.FATAL, tag, msg, exception)
        }

        Log.removeLogger(logger)
    }
}


lib/src/test/kotlin/me/xizzhu/android/logger/FileLoggerTest.kt
/*
 * Copyright (C) 2021 Xizhi Zhu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xizzhu.android.logger

import java.io.File
import java.text.SimpleDateFormat
import java.util.*
import kotlin.test.*

class FileLoggerTest {
    private lateinit var logFile: File
    private lateinit var logger: FileLogger

    private val tag = "TAG"
    private val msg = "msg"
    private val exception = RuntimeException("Random exception")

    @BeforeTest
    fun setup() {
        logFile = File.createTempFile("test_", ".log")
        logger = FileLogger(logFile = logFile, executor = { it.run() })
    }

    @Test
    fun testLog() {
        logger.log(Log.VERBOSE, tag, msg, null)
        logger.log(Log.DEBUG, tag, msg, null)
        logger.log(Log.INFO, tag, msg, null)
        logger.log(Log.WARN, tag, msg, null)
        logger.log(Log.ERROR, tag, msg, null)
        logger.log(Log.FATAL, tag, msg, null)

        logFile.forEachLineIndexed { index, line ->
            // make sure the first part is a valid timestamp
            SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.US).parse(line.substring(0, 23))

            val level = when (index) {
                0 -> 'V'
                1 -> 'D'
                2 -> 'I'
                3 -> 'W'
                4 -> 'E'
                5 -> 'F'
                else -> fail()
            }
            assertTrue(line.endsWith(" $level/TAG: msg"))
        }
    }

    @Test
    fun testLogWithException() {
        logger.log(Log.VERBOSE, tag, msg, exception)
        logger.log(Log.DEBUG, tag, msg, exception)
        logger.log(Log.INFO, tag, msg, exception)
        logger.log(Log.WARN, tag, msg, exception)
        logger.log(Log.ERROR, tag, msg, exception)
        logger.log(Log.FATAL, tag, msg, exception)

        logFile.forEachLineIndexed { index, line ->
            if (index % 2 != 0) {
                assertEquals(exception.message, line)
                return@forEachLineIndexed
            }

            // make sure the first part is a valid timestamp
            SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.US).parse(line.substring(0, 23))

            val level = when (index) {
                0 -> 'V'
                2 -> 'D'
                4 -> 'I'
                6 -> 'W'
                8 -> 'E'
                10 -> 'F'
                else -> fail()
            }
            assertTrue(line.endsWith(" $level/TAG: msg"))
        }
    }

    private fun File.forEachLineIndexed(action: (index: Int, line: String) -> Unit) {
        var index = 0
        forEachLine { action(index++, it) }
    }
}


lib/src/main/kotlin/me/xizzhu/android/logger/Logger.kt
/*
 * Copyright (C) 2021 Xizhi Zhu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xizzhu.android.logger

import androidx.annotation.IntDef
import java.util.concurrent.CopyOnWriteArrayList

interface Logger {
    @Log.Level
    var level: Int

    fun log(@Log.Level level: Int, tag: String, msg: String, e: Throwable?)
}

object Log {
    const val VERBOSE = 1
    const val DEBUG = 2
    const val INFO = 3
    const val WARN = 4
    const val ERROR = 5
    const val FATAL = 6

    @IntDef(VERBOSE, DEBUG, INFO, WARN, ERROR, FATAL)
    @Retention(AnnotationRetention.SOURCE)
    annotation class Level

    private val loggers = CopyOnWriteArrayList<Logger>()

    fun addLogger(logger: Logger) {
        loggers.addIfAbsent(logger)
    }

    fun removeLogger(logger: Logger) {
        loggers.remove(logger)
    }

    fun v(tag: String, msg: String) {
        log(VERBOSE, tag, msg)
    }

    fun v(tag: String, msg: String, e: Throwable) {
        log(VERBOSE, tag, msg, e)
    }

    fun d(tag: String, msg: String) {
        log(DEBUG, tag, msg)
    }

    fun d(tag: String, msg: String, e: Throwable) {
        log(DEBUG, tag, msg, e)
    }

    fun i(tag: String, msg: String) {
        log(INFO, tag, msg)
    }

    fun i(tag: String, msg: String, e: Throwable) {
        log(INFO, tag, msg, e)
    }

    fun w(tag: String, msg: String) {
        log(WARN, tag, msg)
    }

    fun w(tag: String, msg: String, e: Throwable) {
        log(WARN, tag, msg, e)
    }

    fun e(tag: String, msg: String) {
        log(ERROR, tag, msg)
    }

    fun e(tag: String, msg: String, e: Throwable) {
        log(ERROR, tag, msg, e)
    }

    fun f(tag: String, msg: String) {
        log(FATAL, tag, msg)
    }

    fun f(tag: String, msg: String, e: Throwable) {
        log(FATAL, tag, msg, e)
    }

    private fun log(@Level level: Int, tag: String, msg: String, e: Throwable? = null) {
        for (logger in loggers) {
            if (level >= logger.level) {
                logger.log(level, tag, msg, e)
            }
        }
    }
}


app/src/main/kotlin/me/xizzhu/android/logger/sample/App.kt
/*
 * Copyright (C) 2021 Xizhi Zhu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xizzhu.android.logger.sample

import android.app.Application
import me.xizzhu.android.logger.FileLogger
import me.xizzhu.android.logger.Log
import me.xizzhu.android.logger.LogcatLogger
import java.io.File

class App : Application() {
    override fun onCreate() {
        super.onCreate()

        if (BuildConfig.DEBUG) {
            Log.addLogger(LogcatLogger())
        }
        Log.addLogger(FileLogger(File(filesDir, "xizzhu.log")))
    }
}


lib/src/main/kotlin/me/xizzhu/android/logger/FileLogger.kt
/*
 * Copyright (C) 2021 Xizhi Zhu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xizzhu.android.logger

import java.io.File
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.Executor
import java.util.concurrent.Executors

class FileLogger(logFile: File,
                 @Log.Level override var level: Int = Log.VERBOSE,
                 private val executor: Executor = Executors.newSingleThreadExecutor()) : Logger {
    companion object {
        private val TAG = FileLogger::class.java.simpleName
    }

    private val fos: FileOutputStream by lazy { FileOutputStream(logFile, true) }
    private val dateFormat: SimpleDateFormat by lazy { SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.US) }

    override fun log(@Log.Level level: Int, tag: String, msg: String, e: Throwable?) {
        executor.execute {
            try {
                synchronized(fos) {
                    fos.write(formatLog(level, tag, msg, e).toByteArray())
                    fos.flush()
                }
            } catch (e: Exception) {
                android.util.Log.e(TAG, "Failed to log to file", e)
            }
        }
    }

    private fun formatLog(@Log.Level level: Int, tag: String, msg: String, e: Throwable?): String {
        val sb = StringBuilder()
                .append(dateFormat.format(Date()))
                .append(' ')
                .append(when (level) {
                    Log.VERBOSE -> 'V'
                    Log.DEBUG -> 'D'
                    Log.INFO -> 'I'
                    Log.WARN -> 'W'
                    Log.ERROR -> 'E'
                    Log.FATAL -> 'F'
                    else -> throw IllegalArgumentException("Unsupported log level - $this")
                })
                .append('/')
                .append(tag)
                .append(": ")
                .append(msg)
                .append('\n')
        e?.let { sb.append(android.util.Log.getStackTraceString(it)).append('\n') }
        return sb.toString()
    }
}


lib/src/main/kotlin/me/xizzhu/android/logger/LogcatLogger.kt
/*
 * Copyright (C) 2021 Xizhi Zhu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xizzhu.android.logger

class LogcatLogger(@Log.Level override var level: Int = Log.VERBOSE) : Logger {
    override fun log(@Log.Level level: Int, tag: String, msg: String, e: Throwable?) {
        if (e != null) {
            when (level) {
                Log.VERBOSE -> android.util.Log.v(tag, msg, e)
                Log.DEBUG -> android.util.Log.d(tag, msg, e)
                Log.INFO -> android.util.Log.i(tag, msg, e)
                Log.WARN -> android.util.Log.w(tag, msg, e)
                Log.ERROR -> android.util.Log.e(tag, msg, e)
                Log.FATAL -> android.util.Log.wtf(tag, msg, e)
            }
        } else {
            when (level) {
                Log.VERBOSE -> android.util.Log.v(tag, msg)
                Log.DEBUG -> android.util.Log.d(tag, msg)
                Log.INFO -> android.util.Log.i(tag, msg)
                Log.WARN -> android.util.Log.w(tag, msg)
                Log.ERROR -> android.util.Log.e(tag, msg)
                Log.FATAL -> android.util.Log.wtf(tag, msg)
            }
        }
    }
}


lib/src/main/kotlin/me/xizzhu/android/logger/CrashlyticsLogger.kt
/*
 * Copyright (C) 2021 Xizhi Zhu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xizzhu.android.logger

import com.google.firebase.crashlytics.FirebaseCrashlytics

class CrashlyticsLogger @JvmOverloads constructor(
        @Log.Level override var level: Int = Log.VERBOSE,
        private val crashlytics: FirebaseCrashlytics = FirebaseCrashlytics.getInstance()
) : Logger {
    override fun log(level: Int, tag: String, msg: String, e: Throwable?) {
        if (msg.isNotBlank()) {
            if (tag.isBlank()) {
                crashlytics.log(msg)
            } else {
                crashlytics.log("$tag: $msg")
            }
        }
        e?.let { crashlytics.recordException(it) }
    }
}


app/src/main/kotlin/me/xizzhu/android/logger/sample/MainActivity.kt
/*
 * Copyright (C) 2021 Xizhi Zhu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xizzhu.android.logger.sample

import android.app.Activity
import android.os.Bundle
import me.xizzhu.android.logger.Log

class MainActivity : Activity() {
    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.i(TAG, "onCreate")
        setContentView(R.layout.activity_main)

        Log.e(TAG, "Fake exception", RuntimeException("Just a fake exception"))
    }

    override fun onStart() {
        super.onStart()
        Log.i(TAG, "onStart")
    }

    override fun onStop() {
        super.onStop()
        Log.i(TAG, "onStop")
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.i(TAG, "onDestroy")
    }
}


buildSrc/src/main/kotlin/Configurations.kt
/*
 * Copyright (C) 2021 Xizhi Zhu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import org.gradle.api.JavaVersion

object Configurations {
    val libGroupId = "me.xizzhu"
    val libArtifactId = "logger"
    val sampleAppId = "me.xizzhu.android.logger.sample"
}

object Versions {
    object Logger {
        const val code = 700
        const val name = "${code / 10000}.${(code % 10000) / 100}.${code % 100}"
    }

    object BinaryCompatibilityValidator {
        const val classpath = "0.7.1"
    }

    object Sdk {
        const val classpath = "7.0.3"
        const val buildTools = "31.0.0"
        const val compile = 31
        const val min = 19
        const val target = 31
    }

    val java = JavaVersion.VERSION_11

    object Kotlin {
        const val jvmTarget = "11"
        const val core = "1.5.31"
    }

    object AndroidX {
        const val annotation = "1.2.0"
    }

    object Firebase {
        const val crashlytics = "18.2.3"
    }

    const val mockk = "1.12.0"
}

object Dependencies {
    object Sdk {
        const val classpath = "com.android.tools.build:gradle:${Versions.Sdk.classpath}"
    }

    object BinaryCompatibilityValidator {
        const val classpath = "org.jetbrains.kotlinx:binary-compatibility-validator:${Versions.BinaryCompatibilityValidator.classpath}"
    }

    object Kotlin {
        const val classpath = "org.jetbrains.kotlin:kotlin-gradle-plugin:${Versions.Kotlin.core}"
        const val test = "org.jetbrains.kotlin:kotlin-test-junit:${Versions.Kotlin.core}"
    }

    object AndroidX {
        const val annotation = "androidx.annotation:annotation:${Versions.AndroidX.annotation}"
    }

    object Firebase {
        const val crashlytics = "com.google.firebase:firebase-crashlytics:${Versions.Firebase.crashlytics}"
    }

    object Mockk {
        const val mockk = "io.mockk:mockk:${Versions.mockk}"
    }
}
